### ISSUE TITLE
the global cannot be freezed in vm.runInNewContext in node-v0.12.x

### ISSUE THREAD

**Issue by ddnionio** [joyent/node#25774]
The global cannot be freeze in the vm.runInNewContext in v0.12.x, while it can in v0.10.x

see example: a.js
```javascript
var vm = require("vm");
var code = "(function(){Object.freeze(global); new_var = 2; return global; })"; //the global is freezed, so the new_var should not be added into global

var sandbox = {
	original_var:1
};

sandbox.global = sandbox;
var f=vm.runInNewContext(code, sandbox);
console.log(f());
```

in 0.12.x
node a.js
```
{ original_var: 1, global: [Circular], new_var: 2 } 
```

in 0.10.x
node a.js
```
{ original_var: 1, global: [Circular]}
```

**Comment by jasnell** [comment]
@joyent/node-tsc ... what do you think about this one?

**Comment by domenic** [comment]
Seems likely global vs. global proxy... Will do some testing to confirm and explain if so. Nice find.

**Comment by domenic** [comment]
This actually just appears to be more correct now. You can test in Chrome: `Object.freeze(window); Object.isFrozen(window) === false`.

Whether that's per-spec is not entirely clear to me yet. Firefox throws an error, which is a bit nicer. Edge actually allows freezing. I don't have Safari available to test.

I think https://github.com/domenic/window-proxy-spec is actually the most relevant spec here. It aligns with Firefox throwing. So it'd be a V8 bug that it doesn't throw.

But on the io.js side, I'd say working as expected.

**Comment by jasnell** [comment]
:+1: @domenic ... Possibly worth documenting at least.

**Comment by ddnionio** [comment]
@domenic but I can freeze global in nodejs's startup context (the main-module's context).

What can I do if I want to freeze the global in newContext?

**Comment by domenic** [comment]
@ddnionio that is extremely surprising. Can you post a passing test (in io.js)?

**Comment by ddnionio** [comment]
@domenic I use iojs v2.5.0 for win64
example b.js
```javascript
a=1
Object.freeze(global)
console.log(Object.isFrozen(global));
b=2
console.log(global.a);
console.log(global.b)
```
```
$ iojs b.js
true
1
undefined
```
So it can freeze global in mainmodule. but if i run the a.js in first post, it shows the same result with nodev0.12, the global cannot be frozen in newcontext.

**Comment by domenic** [comment]
Wow. That's a pretty serious bug. We need to fix it so you cannot freeze the global.

**Comment by ddnionio** [comment]
@domenic So it is not allowed to freeze the global, am i right?

**Comment by domenic** [comment]
@ddnionio right. This is kind of at the forefront of standardization, and engines are inconsistent, but there are important security principles behind the decisions in https://github.com/domenic/window-proxy-spec, so eventually all engines will converge on not allowing global freezing.

**Comment by ddnionio** [comment]
@domenic I see, thank you!

**Comment by jasnell** [comment]
@domenic ... would this be a bug in V8 or node? If it's in node, I'm not certain that it's something we'd be able to fix in v0.10 or v0.12. At the very least, we need to document the issue.

**Comment by domenic** [comment]
@jasnell it's hard to say without further investigation. The fact that Chrome already rejects global freezing, but apparently io.js does not, implies maybe a bug on the io.js side. But hard to say.
